rlmolecule/mcts/mcts.py

```diff
@@ -91,19 +93,14 @@ def _select(
         (playout) has yet been initiated. The section below says more about a way of biasing choice of child vertices that
         lets the game tree expand towards the most promising moves, which is the essence of Monte Carlo tree search.
         """
-        current = root
-        search_path = []
-        while True:
-            search_path.append(current)
-            children = current.children
 
-            if children is None:  # node is unexpanded: expand and return its value estimate
-                return search_path, self._evaluate(current, search_path)
+        def _iter_select(parent: MCTSVertex) -> [MCTSVertex]:
+            yield parent
+            if parent.expanded and len(parent.children) > 0:
+                next_parent = max(parent.children, key=lambda child: self._ucb_score(parent, child))
+                yield from _iter_select(next_parent)
```

Comment:
This change is mainly the result of factoring out `_evaluate`

Reply:
MAkes sense. I further simplified the evaluate function to just take the search path.

I also prefer to use loops here as I find them (at least here) to be clearer than a generator. Also, generators and other recursive methods are generally less performant (sometimes the JIT optimizer is able to overcome this, but they start at a disadvantage due to the repeated function calls not present in a loop). For simplicity's sake, I transformed back into a loop implementation.